File: back/game_analysis.py
def analyze_games(games, profile_id):
    profile_id = int(profile_id)
    
    # Initialize statistics dictionaries
    opponents = {}
    allies = {}
    match_stats = {'total': 0, 'wins': 0, 'losses': 0}
    civ_stats = {}
    
    for game in games:
        # Find the player in the teams structure
        player_info = None
        player_team_index = None

        
        for team_index, team in enumerate(game['teams']):
            for team_member in team:
                if team_member['player']['profile_id'] == profile_id:
                    player_info = team_member['player']
                    player_team_index = team_index
                    break
            if player_info:
                break
                
        if not player_info:
            continue
            
        # Update match statistics
        match_stats['total'] += 1
        player_won = player_info['result'] == 'win'
        if player_won:
            match_stats['wins'] += 1
        else:
            match_stats['losses'] += 1
            
        # Update civilization statistics
        player_civ = player_info['civilization']
        if player_civ not in civ_stats:
            civ_stats[player_civ] = {'total': 0, 'wins': 0, 'losses': 0, 'win_rate': 0}
        civ_stats[player_civ]['total'] += 1
        if player_won:
            civ_stats[player_civ]['wins'] += 1
        else:
            civ_stats[player_civ]['losses'] += 1
        civ_stats[player_civ]['win_rate'] = (civ_stats[player_civ]['wins'] / civ_stats[player_civ]['total']) * 100.0
        
        # Process other players (teammates and enemies)
        for team_index, team in enumerate(game['teams']):
            for team_member in team:
                other_player = team_member['player']
                if other_player['profile_id'] == profile_id:
                    continue
                
                name = other_player['name']
                if team_index == player_team_index:
                    if name not in allies:
                        allies[name] = {'games': 0, 'wins': 0, 'losses': 0}
                    allies[name]['games'] += 1
                    if player_won:
                        allies[name]['wins'] += 1
                    else:
                        allies[name]['losses'] += 1
                else:
                    if name not in opponents:
                        opponents[name] = {'games': 0, 'wins': 0, 'losses': 0}
                    opponents[name]['games'] += 1
                    if player_won:
                        opponents[name]['losses'] += 1
                    else:
                        opponents[name]['wins'] += 1
    
    sorted_civ_stats = dict(sorted(
        civ_stats.items(), 
        key=lambda item: item[1]['total'], 
        reverse=True
    ))

    sorted_opponents = dict(sorted(
        opponents.items(), 
        key=lambda item: item[1]['games'], 
        reverse=True
    ))

    sorted_allies = dict(sorted(
        allies.items(), 
        key=lambda item: item[1]['games'], 
        reverse=True
    ))

    
    return {
        'match_stats': match_stats,
        'civ_stats': sorted_civ_stats,
        'allies': sorted_allies,
        'opponents': sorted_opponents
    }

File: back/game_analysis.go
package main

import (
	"sort"
)

func AnalyzeGames(games []Game, profileID int) AnalyzeGamesResult {
	opponents := map[string]*AllyOpponentStat{}
	allies := map[string]*AllyOpponentStat{}
	matchStats := map[string]int{"total": 0, "wins": 0, "losses": 0}
	civStats := map[string]*CivStat{}

	for _, game := range games {
		var playerInfo *Player
		var playerTeamIndex int = -1

		// Find player in teams
		for tIdx, team := range game.Teams {
			for _, member := range team {
				if member.Player.ProfileID == profileID {
					playerInfo = &member.Player
					playerTeamIndex = tIdx
					break
				}
			}
			if playerInfo != nil {
				break
			}
		}
		if playerInfo == nil {
			continue
		}

		// Update match stats
		matchStats["total"]++
		playerWon := playerInfo.Result == "win"
		if playerWon {
			matchStats["wins"]++
		} else {
			matchStats["losses"]++
		}

		// Update civ stats
		pciv := playerInfo.Civilization
		if _, ok := civStats[pciv]; !ok {
			civStats[pciv] = &CivStat{}
		}
		civStats[pciv].Total++
		if playerWon {
			civStats[pciv].Wins++
		} else {
			civStats[pciv].Losses++
		}

		// Process other players
		for tIdx, team := range game.Teams {
			for _, member := range team {
				other := member.Player
				if other.ProfileID == profileID {
					continue
				}
				name := other.Name
				if tIdx == playerTeamIndex {
					if _, ok := allies[name]; !ok {
						allies[name] = &AllyOpponentStat{}
					}
					allies[name].Games++
					if playerWon {
						allies[name].Wins++
					} else {
						allies[name].Losses++
					}
				} else {
					if _, ok := opponents[name]; !ok {
						opponents[name] = &AllyOpponentStat{}
					}
					opponents[name].Games++
					if playerWon {
						opponents[name].Losses++
					} else {
						opponents[name].Wins++
					}
				}
			}
		}
	}

	// Convert civStats to sorted slice then to map (by total desc)
	sortedCivStats := map[string]CivStat{}
	type civPair struct {
		Name string
		Stat *CivStat
	}
	var civPairs []civPair
	for k, v := range civStats {
		civPairs = append(civPairs, civPair{k, v})
	}
	sort.Slice(civPairs, func(i, j int) bool {
		return civPairs[i].Stat.Total > civPairs[j].Stat.Total
	})
	for _, cp := range civPairs {
		sortedCivStats[cp.Name] = *cp.Stat
	}

	// Sort allies and opponents by games desc
	var sortedAllies []NameStatPair
	type aoPair struct {
		Name string
		Stat *AllyOpponentStat
	}
	var allyPairs []aoPair
	for k, v := range allies {
		allyPairs = append(allyPairs, aoPair{k, v})
	}
	sort.Slice(allyPairs, func(i, j int) bool {
		return allyPairs[i].Stat.Games > allyPairs[j].Stat.Games
	})
	for _, ap := range allyPairs {
		sortedAllies = append(sortedAllies, NameStatPair{ap.Name, *ap.Stat})
	}

	var sortedOpponents []NameStatPair
	var oppPairs []aoPair
	for k, v := range opponents {
		oppPairs = append(oppPairs, aoPair{k, v})
	}
	sort.Slice(oppPairs, func(i, j int) bool {
		return oppPairs[i].Stat.Games > oppPairs[j].Stat.Games
	})
	for _, op := range oppPairs {
		sortedOpponents = append(sortedOpponents, NameStatPair{op.Name, *op.Stat})
	}

	return AnalyzeGamesResult{
		MatchStats: matchStats,
		CivStats:   sortedCivStats,
		Allies:     sortedAllies,
		Opponents:  sortedOpponents,
	}
}
